src/aleph/web/controllers/storage.py

```diff
@@ -114,59 +117,168 @@ class StorageMetadata(pydantic.BaseModel):
     sync: bool
 
 
-class UploadedFile(Protocol):
+class UploadedFile:
+    def __enter__(self):
+        raise NotImplementedError
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self.cleanup()
+
+    def cleanup(self):
+        pass
+
+    async def read_and_validate(self):
+        pass
+
     @property
     def size(self) -> int:
-        ...
+        raise NotImplementedError
 
     @property
-    def content(self) -> Union[str, bytes]:
-        ...
+    def content(self) -> bytes:
+        raise NotImplementedError
 
+    @property
+    def file(self):
+        raise NotImplementedError
+
+    def get_hash(self) -> str:
+        raise NotImplementedError
 
-class MultipartUploadedFile:
-    _content: Optional[bytes]
-    size: int
 
-    def __init__(self, file_field: FileField, size: int):
+class MultipartUploadedFile(UploadedFile):
+    def __init__(self, file_field: BodyPartReader, max_size: int):
         self.file_field = file_field
-        self.size = size
-        self._content = None
+        self.max_size = max_size
+        self._hash = None
+        try:
+            self._temp_file = tempfile.NamedTemporaryFile(delete=False)
+            self._file_content = bytearray()
+        except Exception as e:
+            web.HTTPInternalServerError(reason="Cannot create tempfile")
+
+    def __enter__(self):
+        self._temp_file.seek(0)
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        try:
+            self._temp_file.close()
+            if os.path.exists(self._temp_file.name):
+                os.unlink(self._temp_file.name)
+        except Exception as e:
+            web.HTTPInternalServerError(reason="Cannot create tempfile")
+
+    async def read_and_validate(self):
+        total_read = 0
+        chunk_size = 8192
+        hash_sha256 = hashlib.sha256()
+
+        while total_read < self.max_size:
+            chunk = await self.file_field.read_chunk(chunk_size)
+            if not chunk:
+                break
+            total_read += len(chunk)
+            if total_read > self.max_size:
+                raise web.HTTPRequestEntityTooLarge(
+                    reason="File size exceeds the maximum limit.",
+                    max_size=self.max_size,
+                    actual_size=total_read,
+                )
+            self._temp_file.write(chunk)
+            self._file_content.extend(chunk)
+            hash_sha256.update(chunk)
+        self._hash = hash_sha256.hexdigest()
+        self._temp_file.seek(0)
```

Comment:
Is there a reason on why you do that? You probably want to add a comment here explaining this because it doesn't seem obvious to me so I guess other people won't guess it neither.